Acceleration/deceleration

ElementalAlly opened the pull request:
Acceleration ramping with separation through classes. This is tested and working.

vladaviedov committed:
Merge pull request #34 from recordrobotics/dev

Subsystems Complete

ElementalAlly committed:
First Working Build (Acceleration Ramping)

This is the first buildable version, only with the acceleration ramping.

ElementalAlly committed:
Acceleration Ramping on Main Robot Code

Ported acceleration ramping from testbed to actual program. Will test
soon.

ElementalAlly committed:
Moved Acceleration Ramping to Separate Class

Previously, acceleration ramping had been a part of the drive command.
Now, it is in a separate file, which can be reused to implement
acceleration/deceleration ramping elsewhere.

ElementalAlly commented:
I don't know how the heck to resolve the conflicts in RobotContainer...

fielia commented on src/main/java/org/recordrobotics/munchkin/Constants.java:
```diff
@@ -4,5 +4,9 @@
+	public final static double RAMPING_JUMP_THRESHOLD = 0.1;
+	public final static double DRIVE_CONTROL_RAMPING = 25.0;
+	public final static double DRIVE_MAX_CONTROL_RAMPING = 15.0;
 }
```
Do all these constants need to be here? Zero, for example, feels unnecessary
also, are these values going to be used anywhere except manualdrive? this file is for constants that will be used in multiple files. otherwise, just have them declared locally

vladaviedov replied:
I second this.
ZERO is also a poor name choice, it's no better than leaving the magic number in the code, give it a name for what it actually represents.

fielia replied:
As an example, COMMANDS_TAB and DATA_TAB are pretty specific, as there's only one thing they could be referring to (shuffleboard). They also are/could be used in many files, like Climbers, Rotator, and Sensors. These types of constants go in Constants.java.

fielia commented on src/main/java/org/recordrobotics/munchkin/commands/manual/ManualDrive.java:
```diff
@@ -14,8 +16,12 @@ public class ManualDrive extends CommandBase {
 	private IControlInput _controls;
+	private double _controlScaleLong;
+	private double _controlScaleLat;
+	private Boolean _rampingBool;
```
could this name be more descriptive? i think something like _ramp or _isRamping is better

fielia replied:
I think in general, most of your names could be more descriptive (ex: i have no idea what the drive control ramping vars are for
an alternative for this, if u can't think up better names, is to use javadoc comments to describe the variables (here, u can use sentences lol)

fielia commented on src/main/java/org/recordrobotics/munchkin/control/ControlRamping.java:
```diff
@@ -0,0 +1,37 @@
+			if(input == Constants.ZERO) {
+				timeScale = _higherTimeScale;
+			}
+		}
```
this doesn't need to be a nested if
in retrospect, this may be for reading purposes, in which case that should be fine

vladaviedov replied:
This whole construction can be converted into a single 'if' statement

fielia commented on src/main/java/org/recordrobotics/munchkin/control/ControlRamping.java:
```diff
@@ -0,0 +1,37 @@
+	public ControlRamping (double higherTimeScale, double lowerTimeScale, boolean maxScaling) {
+		_higherTimeScale = higherTimeScale;
+		_lowerTimeScale = lowerTimeScale;
+		_maxScaling = maxScaling;
```
This may just be me being dumb, but what's the point of maxScaling? what does it do?

ElementalAlly replied:
Maxscaling enables the acceleration to scale faster if the control stick is pushed all the way forwards/backwards.

fielia replied:
use javadocs to explain that in code

fielia commented on src/main/java/org/recordrobotics/munchkin/control/ControlRamping.java:
```diff
@@ -0,0 +1,37 @@
+		_maxScaling = maxScaling;
+	}
+
+	public double calcNextCtrlScale(double currControlScale, double input) {
```
what does this function do? adding javadoc comments here, to ur contructor, and to ur variables should help with readability (and help us not have to decipher ur entire code when calling the function again)

vladaviedov replied:
Yes, this function definitely needs some comments, it's very hard to understand.

the-goat-man reviewed (changes requested):
I'm prolly just being nitpicky, but i think ur concepts and executions are great!
also, i may have missed some stuff here

ElementalAlly committed:
Merge branch 'dev' into accelerationDeceleration

vladaviedov commented on build.gradle:
```diff
@@ -11,7 +11,7 @@ targetCompatibility = JavaVersion.VERSION_11
 
 // PMD config
-pmd {
+ pmd {
```
Revert file, please

vladaviedov commented on src/main/java/org/recordrobotics/munchkin/RobotContainer.java:
```diff
@@ -123,6 +123,12 @@ private void controlDouble() {
+			new ManualClimbers(_climbers, _controlInput),
+			new ManualFlywheel(_flywheel, _controlInput),
+			new ManualRotator(_rotator, _controlInput),
+			new ManualDrive(_drive, _controlInput, true));
```
This should not be here.
See initTeleopCommands() for the proper way to create commands.
Note that the for loop schedules all commands inside _teleopPairs

vladaviedov replied:
I realize this was probably from the merge conflict, but it still needs to be changed

fielia replied:
I think Vlad is pointing at the third parameter in the ManualDrive() call. The issue is it doesn't align with the call in initTeleopCommands().

vladaviedov commented on src/main/java/org/recordrobotics/munchkin/control/ControlRamping.java:
```diff
@@ -0,0 +1,37 @@
+
+import org.recordrobotics.munchkin.Constants;
+
+public class ControlRamping {
```
First, the 'control' directory is meant for abstracting away the interaction between input devices and robot code.
Second, this class should really be a function, it doesn't make much sense to allocate an object for it.

ElementalAlly replied:
Here, I make it a separate class so it can be used throughout the code in different subsystems, instead of having to manually copy the code to each :) .

vladaviedov commented on src/main/java/org/recordrobotics/munchkin/commands/manual/ManualDrive.java:
```diff
@@ -17,8 +19,12 @@ public class ManualDrive extends CommandBase {
+	private ControlRamping _ramper;
 
-	public ManualDrive(Drive drive, IControlInput controls) {
+	public ManualDrive(Drive drive, IControlInput controls, Boolean rampingBool) {
```
This actually introduces a possible NPE with no benefit.
Use the primitive version of boolean (lowercase) instead.

fielia replied:
for the people, an NPE is a Null Pointer Exception

vladaviedov commented on src/main/java/org/recordrobotics/munchkin/control/ControlRamping.java:
```diff
@@ -0,0 +1,37 @@
+		if(Math.abs(input-currControlScale) <= Constants.RAMPING_JUMP_THRESHOLD) {
+			nextCtrlScale = input;
+		}
+		if(nextCtrlScale < -1.0 || nextCtrlScale > 1.0) {
```
Replace with Math.abs call

vladaviedov reviewed (changes requested):
Messy; needs documentation

the-goat-man committed:
reverting build.gradle changes

fielia commented on src/main/java/org/recordrobotics/munchkin/commands/manual/ManualDrive.java:
```diff
@@ -17,8 +19,12 @@ public class ManualDrive extends CommandBase {
+	private ControlRamping _ramper;
 
-	public ManualDrive(Drive drive, IControlInput controls) {
+	public ManualDrive(Drive drive, IControlInput controls, Boolean rampingBool) {
```
I'm curious why you decided to import the value of rampingBool, instead of having it set as a constant. The only reason I can think of to change whether we ramp the accel/decel (if that's what _rampingBool is supposed to do) in the same program call is if we want to have a shuffleboard button switch whether we ramp or not. However, there is a much better way of doing this than adding a parameter to the constructor.

ElementalAlly replied:
I don't know how to interface with shuffleboard, and if you could help me with that (that activation/deactivation is what I had in mind), that would be great! Thanks!

fielia commented on src/main/java/org/recordrobotics/munchkin/control/ControlRamping.java:
```diff
@@ -0,0 +1,37 @@
+	private double _lowerTimeScale;
+	private boolean _maxScaling;
+
+	public ControlRamping (double higherTimeScale, double lowerTimeScale, boolean maxScaling) {
```
Similar to my comment about the ManualDrive constructor, why have the maxScaling boolean as a parameter, rather than a constant?

ElementalAlly replied:
This is in case I want to use ramping that doesn't accelerate faster when the control stick is all the way forwards, and using this setup, I can change that based on the subsystem the ramper is being used on.

ElementalAlly committed:
Pullable version

Most comments on PR resolved. If you could review my changes and see if
there's any more I should do, that would be great! Thanks so much!

ElementalAlly commented:
I believe I have made all the changes you guys wanted. Thanks for the help and review, and if you could help me or teach me how to interface with Shuffleboard to be able to turn on and off the ramping, that would be great! Thanks so much!

fielia commented on src/main/java/org/recordrobotics/munchkin/Constants.java:
```diff
@@ -4,8 +4,7 @@
  */
 public final class Constants {
-
+	public final static double MOVEMENT_EXPONENTIAL_BASE = 10.0;
```
This variable name still seems vague. It sounds like '10' is supposed to be the x value in x^y, but if this is going to be used in Ramping, i don't think it'll go there

ElementalAlly replied:
artifact from previous version... useless variable

fielia commented on src/main/java/org/recordrobotics/munchkin/RobotContainer.java:
```diff
@@ -67,7 +67,7 @@ private void initTeleopCommands() {
 		_teleopPairs.add(new Pair<Subsystem, Command>(_flywheel, new ManualFlywheel(_flywheel, _controlInput)));
 		_teleopPairs.add(new Pair<Subsystem, Command>(_rotator, new ManualRotator(_rotator, _controlInput)));
-		_teleopPairs.add(new Pair<Subsystem, Command>(_drive, new ManualDrive(_drive, _controlInput)));
+		_teleopPairs.add(new Pair<Subsystem, Command>(_drive, new ManualDrive(_drive, _controlInput, true)));
```
I don't think this was fixed yet. There's still an error in this file, and I'm not even sure the third parameter even belongs

fielia replied:
I was wrong about the error, but I'm still concerned about the existence of the third parameter.

vladaviedov replied:
This is part of ManualDrive constructor, it will not compile otherwise.
I think having ramping as an option that can  be adjusted is perfectly valid.
This way we can toggle it during runtime.

fielia replied:
I agree
I think having it, and ramping in general, be a part of ManualDrive is bad. Instead, include it in Drive, and still be able to toggle it.

fielia replied:
Also, I think the third parameter is unnecessary, and maybe against the point of robot container
I see robot container as holding all the big parts of the robot, like the drive system, climbers, and flywheel, for example, and activating and deactivating those parts at specific times.
From there, the specific parts do different things based on how and when their activated. ideally, this is done independently of the other parts.
ramping only deals with drive, at least for now, and isn't a big part on the robot. thus, i think the logic regarding ramping activation should be localized in drive, but still have a toggle command.

thank you for listening to my ted talk

fielia commented on src/main/java/org/recordrobotics/munchkin/util/ControlRamping.java:
```diff
@@ -0,0 +1,62 @@
+package org.recordrobotics.munchkin.util;
+
+public class ControlRamping {
```
I don't think this was fixed. Instead of util, I think this would best be placed in .commands.manual, as that's where this code would be used. I also think, as Vlad stated earlier, this entire object could be a function in another object.
Also, control files are different in our project. They represent control schemes. I suggest changing your file name.

vladaviedov replied:
The way it's designed currently, this should be a function inside ManualDrive.

Now hypothetically, if this was used in another place, it should be a static function. The three fields are not altered in any way withing the object; it would make perfect sense for them to be either function parameters or constants.

fielia replied:
I think this shouldn't be in ManualDrive at all.
Theoretically, we'd also want to ramp auto code, right? Therefore, ramping should be in Drive. It may be harder to implement shuffleboard and the like, but not too bad.

ElementalAlly replied:
Why would we need ramping on the auto command? The auto drive command should calculate the motor spin values needed to move the robot to those positions, then implement them quickly, no?

vladaviedov replied:
There's an argument to be made either way.
I'm leaning towards to ramping on Auto because complicate things and probably make the commands inaccurate.

fielia replied:
I think having the option to put ramping on auto is better, as it doesn't detract from the manual ramping, but allows ramping to be used in auto if that's ever necessary
Right now, I agree auto drive shouldn't be ramped, but if it's a refactoring difference, leaving the options open is preferable, in my opinion.

fielia commented on src/main/java/org/recordrobotics/munchkin/util/ControlRamping.java:
```diff
@@ -0,0 +1,62 @@
+	 * lowerTimeScale is when it needs to be slower. isMaxScaling Faster determines whether
+	 * the higher scaling value is used when the control stick is pressed all the way forwards,
+	 * instead of just when it returns to a neutral position.
+	 */
```
I don't think this is proper javadoc formatting. I will probably reach out soon and show you how to use javadocs.

fielia commented on src/main/java/org/recordrobotics/munchkin/commands/manual/ManualDrive.java:
```diff
@@ -6,6 +6,8 @@
 import org.recordrobotics.munchkin.control.IControlInput;
 import org.recordrobotics.munchkin.subsystems.Drive;
+import org.recordrobotics.munchkin.util.ControlRamping;
+
```
This new line shouldn't be there

fielia commented on src/main/java/org/recordrobotics/munchkin/util/ControlRamping.java:
```diff
@@ -0,0 +1,62 @@
+	private boolean _isMaxScalingFaster;
+
+	private final static double RAMPING_JUMP_THRESHOLD = 0.1;
+	private final static double CTRL_NEUTRAL_POSITION = 0.0; // If the input is less than this constant, it is considered in a neutral position.
```
I'm going to assume this is a placeholder. If so, the value for our neutral stick position is in the project. I'll have to see if it's applied when finding the stick value, or when moving the motors, but I believe it's the former.
In any case, this constant may not be needed.

fielia commented on src/main/java/org/recordrobotics/munchkin/util/ControlRamping.java:
```diff
@@ -0,0 +1,62 @@
+	@param currControlScale The current control scale, used to determine how far the target is, then how
+	@param input The target control scale, and what the control stick actually inputs.
+	@return nextCtrlScale, based on the distance, and setting of ramping.
+	*/
```
This is a better formatted javadoc, but I have a few concerns. Namely, the length.

fielia replied:
Javadocs are used to show a coder what the function is about, yes, but you seem to expect people to come to this file to read the comment. Instead, javadocs help us see what a function is just by hovering over it (depending on your ide)
this is a scary length of text

vladaviedov replied:
Yeah, the function comment should include a brief description of what the function does (abstractly), which parameters it takes and what it returns.

Like Rit said, IDEs will only pick up the header, so you should implementation details as comment in the body of the function instead.

fielia commented on src/main/java/org/recordrobotics/munchkin/util/ControlRamping.java:
```diff
@@ -0,0 +1,62 @@
+
+	private final static double RAMPING_JUMP_THRESHOLD = 0.1;
+	private final static double CTRL_NEUTRAL_POSITION = 0.0; // If the input is less than this constant, it is considered in a neutral position.
+	private final static double MAX_CTRL_SCALE = 1.0;
```
don't know how i didn't catch this earlier, but the practice is to write static before final. minor change, but noticeable.

fielia commented on src/main/java/org/recordrobotics/munchkin/Constants.java:
```diff
@@ -4,8 +4,7 @@
  */
 public final class Constants {
-
+	public final static double MOVEMENT_EXPONENTIAL_BASE = 10.0;
```
similar to my comment below, this should be written "static final", like the variables below

vladaviedov commented on src/main/java/org/recordrobotics/munchkin/util/ControlRamping.java:
```diff
@@ -0,0 +1,62 @@
+		double timeScale = _lowerTimeScale;
+		boolean isControlNeutral = Math.abs(input) <= CTRL_NEUTRAL_POSITION;
+		boolean isControlMax = Math.abs(input) == MAX_CTRL_SCALE;
+		if (_isMaxScalingFaster && (isControlNeutral || isControlMax) || !_isMaxScalingFaster && isControlNeutral) {
```
Notice that `isControlNeutral` is independent from `isMaxScalingFaster`

This should be equivalent, but more concise
`isControlNeutral || _isMaxScalingFaster && isControlMax`

fielia replied:
I think you have it right now so it makes sense to a viewer without a comment
Vlad's statement works too, but is much easier to calculate for the program. If you're concerned about readability, add a comment!

vladaviedov commented on src/main/java/org/recordrobotics/munchkin/commands/manual/ManualDrive.java:
```diff
@@ -28,13 +36,24 @@ public ManualDrive(Drive drive, IControlInput controls) {
+			_controlScaleLat = _ramper.calcNextCtrlScale(_controlScaleLat, _controls.getDriveLat());
+			_drive.move(_controlScaleLong * SPEED_MODIFIER, _controlScaleLat * SPEED_MODIFIER);
+		}
+		else {
```
Push else back to line 52

fielia replied:
what he's saying:
} else {

vladaviedov reviewed (commented):
Good work, but some things still need to be fixed

vladaviedov commented:
> I believe I have made all the changes you guys wanted. Thanks for the help and review, and if you could help me or teach me how to interface with Shuffleboard to be able to turn on and off the ramping, that would be great! Thanks so much!

Yes, take a look at how switching between control schemes work.

In RobotContainer::initDashCommands, you can see how you would create a new dashboard command
```
tab.add("Double Control", new DashRunProcedure(this::controlDouble));
```

Notice here, the parameters to `add` are `(String, Command)`, which is why you need to use the `DashRunProcedure` wrapper.
The constructor takes a function reference which must be `void()`, so you'd create an new function in `RobotContainer` and pass it as the parameter.

Now writing the actual function is a little complicated.
To access the drive command you'd need to search the `_teleopPairs` list by subsystem and then do a cast into `ManualDrive`.
After that you should be able to figure out what to do.

ElementalAlly committed:
Moved ramping to Drive, fixed Javadoc Comments

the-goat-man commented on src/main/java/org/recordrobotics/munchkin/RobotContainer.java:
```diff
@@ -57,7 +58,7 @@ public RobotContainer() {
 		initEntries();
 		initDashCommands();
-		initTeleopCommands();
+		initTeleopCommands(); // Is it OK to change the order of these? Just want to guarantee that _manualDrive is always init before dashboard command changes a value in it...
```
That shouldn't be an issue. Why do you think it would be an issue?

ElementalAlly replied:
I thought it may have a specific reason...

the-goat-man replied:
I don't see anything that will become a problem either way.

ElementalAlly replied:
I don't know what I was thinking... I did that when I had a previous version that may have required the order...

the-goat-man replied:
that's fine! that's why pull requests are here: so we can double-check each other's work :)

ElementalAlly replied:
yay!

ElementalAlly committed:
Merge branch 'dev' into accelerationDeceleration

ElementalAlly committed:
Fixing Spotless Check

the-goat-man commented on src/main/java/org/recordrobotics/munchkin/commands/manual/ManualDrive.java:
```diff
@@ -28,13 +34,24 @@ public ManualDrive(Drive drive, IControlInput controls) {
 		addRequirements(drive);
 	}
 
+	public void toggleAccRamping() {
```
javadoc comment! doesn't need to be complex, literally just `toggles acceleration ramping` would do.

ElementalAlly committed:
Merge branch 'accelerationDeceleration' of https://github.com/recordrobotics/2022_Rewrite into accelerationDeceleration

the-goat-man commented on src/main/java/org/recordrobotics/munchkin/subsystems/Drive.java:
```diff
@@ -22,6 +26,13 @@ public class Drive extends SubsystemBase {
+	public boolean _isRamping = true;
+
+	private NetworkTableEntry _entryRamping;
 	private MotorControllerGroup _leftMotors = new MotorControllerGroup(_left);
```
i would add a new line between the entry and motors, but it's not a big deal

the-goat-man commented on src/main/java/org/recordrobotics/munchkin/subsystems/Drive.java:
```diff
@@ -33,6 +44,54 @@ public class Drive extends SubsystemBase {
+
+		ShuffleboardTab tab = Shuffleboard.getTab(Constants.DATA_TAB);
+		_entryRamping = tab.add("Drive Ramping", true).getEntry();
+
```
unnecessary new line

the-goat-man commented on src/main/java/org/recordrobotics/munchkin/subsystems/Drive.java:
```diff
@@ -33,6 +44,54 @@ public class Drive extends SubsystemBase {
+	 * @param min minimum value to constrain the input into
+	 * @param max maximum value to constrain the input into
+	 * @return the limited value of the input
+	 */
```
nice javadoc comment!

ElementalAlly replied:
Thanks!

the-goat-man commented on src/main/java/org/recordrobotics/munchkin/subsystems/Drive.java:
```diff
@@ -33,6 +44,54 @@ public class Drive extends SubsystemBase {
+	 * @param max maximum value to constrain the input into
+	 * @return the limited value of the input
+	 */
+	private double limitValue(double value, double min, double max) {
```
is this function necessary? it's only used in one place
you could substitute in the logic into the calc method, and add a small `//` comment above stating why it's there

ElementalAlly replied:
more readable code?

the-goat-man replied:
it's not that big of a deal here. if it were a repeated logic string used, then it works. here, its just in one place, and probably never going to be needed again. i believe adding the `//` comment will have the same effect

the-goat-man commented on src/main/java/org/recordrobotics/munchkin/subsystems/Drive.java:
```diff
@@ -33,6 +44,54 @@ public class Drive extends SubsystemBase {
+		return nextCtrlScale;
+	}
+
+	public void toggleAccRamping() {
```
1. javadoc!
2. why is this in both `ManualDrive.java` and `Drive.java`? if you're only calling the function from Drive, you don't need it in ManualDrive

ElementalAlly replied:
Manual drive is redundant... oops

ElementalAlly commented:
@fielia @vladaviedov Could you review it again? I believe I've fixed your comments... Thanks in advance!

the-goat-man commented on src/main/java/org/recordrobotics/munchkin/subsystems/Drive.java:
```diff
@@ -33,6 +44,54 @@ public class Drive extends SubsystemBase {
+	@Override
+	public void periodic() {
+		_entryRamping.setBoolean(_isRamping);
 		_left[0].getEncoder().setPositionConversionFactor(WHEEL_DIAMETER * Math.PI / GEAR_RATIO);
```
i'd add a new line here too

ElementalAlly replied:
Alright!

ElementalAlly committed:
Clean up of variables and comments

ElementalAlly committed:
Reset ManualDrive in RobotContainer

Since I interface with acceleration ramping through _drive instead of
manualDrive, which I was doing, I changed the drive to a previously
defined drive... Switching it back here :)

the-goat-man commented on src/main/java/org/recordrobotics/munchkin/subsystems/Drive.java:
```diff
@@ -81,17 +70,21 @@ public double calcNextCtrlScale(double _lowerTimeScale, double _higherTimeScale,
 			nextCtrlScale = input;
 		}
-		nextCtrlScale = limitValue(nextCtrlScale, -MAX_CTRL_SCALE, MAX_CTRL_SCALE);
+		nextCtrlScale = Math.min(Math.max(nextCtrlScale, -MAX_CTRL_SCALE), MAX_CTRL_SCALE);
```
what does this min-max calculation do? if you're concerned about readability, adding a small line comment here would help clarify

the-goat-man replied:
also, if you're worried about the value exceeding -1 and 1, there's already a speed limiter that can be applied in `Drive.java`. The function is located in `Subsystem.java`, and one of the other subsystem files has an example of how it's used (`Acquisition.java` maybe?). instead of limiting it here, we can limit the values in Drive

ElementalAlly replied:
oh we already do... oops

the-goat-man commented on src/main/java/org/recordrobotics/munchkin/subsystems/Drive.java:
```diff
@@ -30,35 +30,24 @@ public class Drive extends SubsystemBase {
 	private static final double CTRL_NEUTRAL_POSITION = 0.0; // If the input is less than this constant, it is considered in a neutral position.
 	private static final double MAX_CTRL_SCALE = 1.0;
```
Instead of having MAX_CTRL_SCALE and -MAX_CTRL_SCALE, maybe add another variable called MIN_CTRL_SCALE and set it to -1.0

ElementalAlly committed:
Added comment about limiting Ctrl scale

ElementalAlly commented:
Sorry for the absolutely massive amounts of commits... I'm missing things left and right here :)

ElementalAlly committed:
Removed limiting in calcNextCtrlScale

redundant, we already limit

the-goat-man commented on src/main/java/org/recordrobotics/munchkin/RobotContainer.java:
```diff
@@ -81,6 +81,7 @@ private void initAutoCommand() {
 	private void initDashCommands() {
 		ShuffleboardTab tab = Shuffleboard.getTab(Constants.COMMANDS_TAB);
 		tab.add("Reset Climbers Encoder", new DashResetClimbEncoder(_climbers));
+		tab.add("Drive Acceleration Ramping", new DashRunProcedure(_drive::toggleAccRamping));
```
Instead of having the command name be the variable you're changing, maybe have something like "Toggle Drive Ramping"

ElementalAlly replied:
Just fixed this with the new commit

the-goat-man commented on src/main/java/org/recordrobotics/munchkin/subsystems/Drive.java:
```diff
@@ -22,17 +26,64 @@ public class Drive extends SubsystemBase {
+	 * Calculates the next input value for ramping.
+	 * @param _lowerTimeScale slower ramping time scale
+	 * @param _higherTimeScale faster ramping time scale
+	 * @param _isMaxScalingFaster if you push the joystick to its max, this determines whether it ramps faster
```
may be a bit nitpicky, but this can be reworded better
"determines whether pushing joystick to its max ramps faster" or something

ElementalAlly replied:
Fixed this with the new commit

the-goat-man reviewed (dismissed):
it's good to go! other than a few small description changes, ur good! nice work

ElementalAlly committed:
Reworded some names and comments

the-goat-man commented on src/main/java/org/recordrobotics/munchkin/subsystems/Drive.java:
```diff
@@ -22,17 +26,64 @@ public class Drive extends SubsystemBase {
 		new CANSparkMax(RobotMap.DriveBase.RIGHT_BACK_MOTOR_PORT, CANSparkMax.MotorType.kBrushless)
 	};
 
+	private static final double RAMPING_JUMP_THRESHOLD = 0.1;
```
just curious what this represents

ElementalAlly replied:
The threshold between the target control scale and the current control scale that ramping stops and the current is set to the target.

the-goat-man replied:
could you add that in a comment? i think that's all you need
also, ur wording is a bit confusing. `RAMPING_JUMP_THRESHOLD` is the minimum difference allowed between the target scale and current scale?

ElementalAlly replied:
If the distance is less than the threshold, the current will "jump" to the target

the-goat-man replied:
basically, the minimum for having another iteration of the function
just say that! just put ur comment as the javadoc comment

ElementalAlly replied:
Is there any way I can just make this change without you and vlad having to rereview it?

the-goat-man replied:
i don't think so
we don't mind tho

ElementalAlly committed:
Trivial comment add

ElementalAlly committed:
Merge branch 'accelerationDeceleration' of https://github.com/recordrobotics/2022_Rewrite into accelerationDeceleration

ElementalAlly commented:
Sorry for the trouble, but I added a small comment... Could you just rereview this quickly? Thanks!
